### PR TITLE
input: fix stack overflow

### LINKED ISSUE #12287: Alt + cyrillic Er causes stack smashing crash
- `nvim --version`:
NVIM v0.4.3
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.4.3/src -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include

- `vim -u DEFAULTS` does not crash; version:
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Apr 26 2020 11:23:21)
Included patches: 1-510

- Operating system/version:
Linux 5.4.35-1-MANJARO
- Terminal name/version:
xfce4-terminal 0.8.9.2
- `$TERM`:
xterm-256color

### Steps to reproduce using `nvim -u NORC`
```
nvim -u NORC
```
- Enter insert mode
- Press Alt-р (that is Cyrillic "er")

### Actual behaviour
*** stack smashing detected ***: terminated
~                                          [1]    261461 abort (core dumped)  nvim -u NORC
### Expected behaviour
No crash

### PR DESCRIPTION
**Pull request by erw7**
fixes #12287, #11788

### CODE REVIEW

**Comment by teto** [comment]
cc @OVlasiuk @YaLTeR does it fix your problem ?

**Comment by erw7** [comment]
Please squash https://github.com/neovim/neovim/pull/12376/commits/e01fc33515d5cd3a9a595d0e6905bd16d3a861e5 before merging.

**Comment by YaLTeR** [comment]
Can't help unfortunately because I haven't seen my crash in a while. May be related to me switching distributions and DEs.

**Comment by OVlasiuk** [comment]
> cc @OVlasiuk @YaLTeR does it fix your problem ?
Yes, looks good to me. Thanks @erw7 !

File: src/nvim/os/input.c [hunk 1]
```diff
@@ -188,8 +188,13 @@ size_t input_enqueue(String keys)
...
 
-  while (rbuffer_space(input_buffer) >= 6 && ptr < end) {
-    uint8_t buf[6] = { 0 };
+  while (rbuffer_space(input_buffer) >= 19 && ptr < end) {
+    // A "<x>" form occupies at least 1 characters, and produces up
+    // to 19 characters (1 + 5 * 3 for the char and 3 for a modifier).
+    // In the case of K_SPECIAL(0x80) or CSI(0x9B), 3 bytes are escaped and
+    // needed, but since the keys are UTF-8, so the first byte cannot be
+    // K_SPECIAL(0x80) or CSI(0x9B).
+    uint8_t buf[19] = { 0 };
```
Comment (teto) [inline src/nvim/os/input.c:197]: I believe we should use MB_MAXBYTES + 1 as that's mandated by `add_char2buf`. You seem to know a lot about the encoding stuff and I fail at understanding in what situation we can have ` one 16-bit character of up to three bytes plus six following composing characters of three bytes each.` at https://github.com/neovim/neovim/blob/fc98f2d5815d8f7e671766db255edbb7365960b1/src/nvim/vim.h#L287 do you have any idea ?

Comment (erw7) [reply src/nvim/os/input.c:197]: > You seem to know a lot about the encoding stuff and I fail at understanding in what situation we can have one 16-bit character of up to three bytes plus six following composing characters of three bytes each. at ... do you have any idea ?

I'm not particularly familiar with encoding. I looked at the other parts of using MB_MAXBYTES, but I couldn't see when the situation would be similar to the one in the comment.

Comment (erw7) [reply src/nvim/os/input.c:197]: > I believe we should use `MB_MAXBYTES + 1` as that's mandated by `add_char2buf`.

I think the comment on `add_char2buf` is wrong. `add_char2buf` allocates a buffer of `MB_MAXBYTES+1`, but `utf_char2bytes` only uses a maximum of 6 bytes. The first byte is never `K_SPECIAL(0x80)`, so the maximum buffer needed by `add_char2buf` is `3 * 5 + 1 = 16` bytes.

Comment (teto) [reply src/nvim/os/input.c:197]: Agreed.

**Pull request merged**